src/shared/components/templates-grid/TemplatesGrid.tsx

```diff
@@ -0,0 +1,65 @@
+import React, { useMemo, useCallback } from 'react';
+import { useHistory } from 'react-router';
+
+import { Spaceholder } from 'ui';
+
+import { Template } from 'core/api';
+
+import { TemplateTile } from '..';
+
+import csx from './TemplatesGrid.scss';
+
+namespace TemplatesGrid {
+  export interface Props {
+    loading: boolean;
+    templates: Template[];
+    spaceholdersCount?: number;
+    path?: string;
+  }
+}
+
+const renderSpaceholders = (loading: boolean, spaceholdersCount: number) => () =>
+  loading
+    ? Array.from({ length: spaceholdersCount }, (_, idx) => (
+        <Spaceholder key={idx} className={csx.spaceholder} />
+      ))
+    : null;
+
+/** 
+ * Renders grid components based on given data
+ * 
+ * @param props
+ * @param props.loading Displays spaceholders if truthy
+ * @param props.templates Templates to display
+ * @param props.path Path prefix to rediect
+ * @param props.spaceholdersCount Numbers of spaceholders to render when `loading` is truthy
+ */
+const TemplatesGrid = ({
+  loading,
+  templates,
+  path,
+  spaceholdersCount = 12
+}: TemplatesGrid.Props) => {
+  const { push, location } = useHistory();
+
+  const goToDetails = useCallback(
```

Comment:
Adrianku czym się różni to od braku zastosowania useCallback, czyli
const goToDefault = () => { 
...code
}?

Reply:
Wyjasnione przegadane na callu

Reply:
Jeżeli się nie mylę i dobrze to rozumiem (do rewizji przez @polubis ), to w przypadku gdy jest useCallback on nie tworzy nowej funkcji za każdym razem w przypadku re-renderu, tylko w momencie gdy zmieni się to co jest w tej tablicy (czyli location.pathname). W przeciwnym przypadku tworzył by nową funkcje za każdym razem (za każdym renderem), to taki zabieg optymalizacyjny - tak ja to przynajmniej rozumiem. 
Adrian popraw mnie jeśli gadam głupoty :D